package/MDAnalysis/analysis/base.py

```diff
@@ -430,25 +471,231 @@ def run(self, start=None, stop=None, step=None, frames=None,
         logger.info("Starting preparation")
-        self._prepare()
         logger.info("Starting analysis loop over %d trajectory frames",
                     self.n_frames)
 
-        for i, ts in enumerate(ProgressBar(
-                self._sliced_trajectory,
+        frame_indices, trajectory = self._bslices[bslice_idx]
+        for idx, ts in enumerate(ProgressBar(trajectory,
                 verbose=verbose,
                 **progressbar_kwargs)):
+            i = frame_indices[idx]
             self._frame_index = i
             self._ts = ts
             self.frames[i] = ts.frame
             self.times[i] = ts.time
             self._single_frame()
         logger.info("Finishing up")
+        return self
+
+    def _setup_bslices(self, start=None, stop=None, step=None, frames=None): 
+        """
+        Set the self._bslices (the workload distribution scheme) for the future delayed
+        computations.
+
+        Parameters
+        ----------
+        start : int, optional
+            start frame of analysis
+        stop : int, optional
+            stop frame of analysis
+        step : int, optional
+            number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory; `frames` can
+            only be used *instead* of `start`, `stop`, and `step`. Setting
+            *both* `frames` and at least one of `start`, `stop`, `step` to a
+            non-default value will raise a :exc:`ValueError`.
+
+        Raises
+        ------
+        ValueError
+            if *both* `frames` and at least one of `start`, `stop`, or `frames`
+            is provided (i.e., set to another value than ``None``)
+
+        Returns
+        -------
+        bslices : list of (bstart, bstop, bstep, bframes) tuples.
+            Iterator will have size of self._n_bslices.
+        """
+        if frames is not None:
+            if not all(opt is None for opt in [start, stop, step]):
+                raise ValueError("start/stop/step cannot be combined with "
+                                 "frames")
+
+        else:  # frames is None
+            start, stop, step = self._trajectory.check_slice_indices(start, stop, step)
+            frames = list(range(start, stop, step))
+        
+        self.n_frames_total = len(frames)
+        n_bslices = self._n_bslices
+
+        def array_is_boolean(arr):
+            if isinstance(arr, np.ndarray) and arr.dtype is bool:
+                return True
+            elif set(arr) == {True, False}:
+                return True
+            return False
+
+        if array_is_boolean(frames):
+            frames = np.arange(self.n_frames_total)[frames]
+
+        slices = [(indices, self._trajectory[frames_]) 
+                   for indices, frames_ in zip(
+                    np.array_split(range(len(frames)), n_bslices),
+                    np.array_split(frames, n_bslices),
```

Comment:
Would defining this before `slices=` make the code a bit more readable? You could also do a small optimization in the case `array_is_boolean`, since `frames` and `range(len(frames))` would be the same?

Would using NumPy be more efficient to define the range?

Reply:
This whole part is slightly re-written now, and hopefully slightly optimized -- I moved the `enumerate` part to numpy completely.